### PR TITLE
#6833 user ns: Fix segfault while constructing id mappings

### LINKED ISSUE #6833: CRI-O 1.26 / 1.27 panics while starting containers in user namespace on Kubernetes 1.27
### What happened?

If Kubernetes 1.27 tries to start a pod with `hostUsers: false` containers should be started in a user namespace.  

Starting these containers results in the following panic in CRI-O 1.26 and 1.27:
```
crio[2286]: time="2023-04-20 11:34:35.054129181+02:00" level=info msg="Creating container: aquasecurity/trivy-operator-6fcd987f96-vcl9t/trivy-operator" id=09dd0bd5-f359-420c-af2f-d4cd9679c49b name=/runtime.v1.RuntimeService/CreateContainer
crio[2286]: time="2023-04-20 11:34:35.054470669+02:00" level=warning msg="Allowed annotations are specified for workload [io.containers.trace-syscall]"
crio[2286]: panic: runtime error: index out of range [0] with length 0
crio[2286]: goroutine 290 [running]:
crio[2286]: panic({0x560488f072c0, 0xc000acc438})
crio[2286]:         /usr/lib/go/src/runtime/panic.go:987 +0x3bb fp=0xc0011d3b88 sp=0xc0011d3ac8 pc=0x5604870d871b
crio[2286]: runtime.goPanicIndex(0x0, 0x0)
crio[2286]:         /usr/lib/go/src/runtime/panic.go:113 +0x7f fp=0xc0011d3bc8 sp=0xc0011d3b88 pc=0x5604870d66df
crio[2286]: github.com/cri-o/cri-o/server.getOCIMappings(...)
crio[2286]:         /build/cri-o/src/cri-o-1.27.0/server/container_create_linux.go:1083
crio[2286]: github.com/cri-o/cri-o/server.addOCIBindMounts({0x5604890468c8?, 0xc00118fa10?}, {0x56048905d040, 0xc001192480}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0, ...}, ...)
crio[2286]:         /build/cri-o/src/cri-o-1.27.0/server/container_create_linux.go:1053 +0x1d10 fp=0xc0011d4198 sp=0xc0011d3bc8 pc=0x56048866a990
crio[2286]: github.com/cri-o/cri-o/server.(*Server).createSandboxContainer(0xc0000af500, {0x5604890468c8?, 0xc00118f4a0?}, {0x56048905d040, 0xc001192480}, 0xc000253500)
crio[2286]:         /build/cri-o/src/cri-o-1.27.0/server/container_create_linux.go:326 +0x1625 fp=0xc0011d5640 sp=0xc0011d4198 pc=0x5604886614c5
crio[2286]: github.com/cri-o/cri-o/server.(*Server).CreateContainer(0xc0000af500, {0x5604890468c8, 0xc00118f4a0}, 0xc00118e810)
crio[2286]:         /build/cri-o/src/cri-o-1.27.0/server/container_create.go:417 +0xe7c fp=0xc0011d58f8 sp=0xc0011d5640 pc=0x56048865db5c
crio[2286]: k8s.io/cri-api/pkg/apis/runtime/v1._RuntimeService_CreateContainer_Handler.func1({0x5604890468c8, 0xc00118f4a0}, {0x560488f848c0?, 0xc00118e810})
crio[2286]:         /build/cri-o/src/pkg/mod/k8s.io/kubernetes/staging/src/k8s.io/cri-api@v0.0.0-20230411170423-1b4df30b3cdf/pkg/apis/runtime/v1/api.pb.go:10845 +0x78 fp=0xc0011d5938 sp=0xc0011d58f8 pc=0x560487c9edd8
crio[2286]: github.com/cri-o/cri-o/server/otel-collector.UnaryInterceptor.func1({0x5604890468c8, 0xc00118e7e0}, {0x560488f848c0, 0xc00118e810}, 0xc000a8c580, 0xc000946150)
crio[2286]:         /build/cri-o/src/cri-o-1.27.0/server/otel-collector/interceptors.go:64 +0x1a6 fp=0xc0011d5a10 sp=0xc0011d5938 pc=0x5604886ae0a6
crio[2286]: k8s.io/cri-api/pkg/apis/runtime/v1._RuntimeService_CreateContainer_Handler({0x56048901cdc0?, 0xc0000af500}, {0x5604890468c8, 0xc00118e7e0}, 0xc0003720e0, 0x560489021ac8)
crio[2286]:         /build/cri-o/src/pkg/mod/k8s.io/kubernetes/staging/src/k8s.io/cri-api@v0.0.0-20230411170423-1b4df30b3cdf/pkg/apis/runtime/v1/api.pb.go:10847 +0x138 fp=0xc0011d5a68 sp=0xc0011d5a10 pc=0x560487c9ec98
crio[2286]: google.golang.org/grpc.(*Server).processUnaryRPC(0xc000036000, {0x56048904f798, 0xc000e54000}, 0xc000993680, 0xc000891dd0, 0x560489ec23b0, 0x0)
crio[2286]:         /build/cri-o/src/pkg/mod/google.golang.org/grpc@v1.54.0/server.go:1345 +0xdf3 fp=0xc0011d5e48 sp=0xc0011d5a68 pc=0x560487c64533
crio[2286]: google.golang.org/grpc.(*Server).handleStream(0xc000036000, {0x56048904f798, 0xc000e54000}, 0xc000993680, 0x0)
crio[2286]:         /build/cri-o/src/pkg/mod/google.golang.org/grpc@v1.54.0/server.go:1722 +0xa36 fp=0xc0011d5f68 sp=0xc0011d5e48 pc=0x560487c69656
crio[2286]: google.golang.org/grpc.(*Server).serveStreams.func1.2()
crio[2286]:         /build/cri-o/src/pkg/mod/google.golang.org/grpc@v1.54.0/server.go:966 +0x98 fp=0xc0011d5fe0 sp=0xc0011d5f68 pc=0x560487c61e58
crio[2286]: runtime.goexit()
[...]
```
(truncated because of GitHub character limit)

### What did you expect to happen?

The container should be successfully started in a user namespace without panicking.

### How can we reproduce it (as minimally and precisely as possible)?

Deploy a Pod with `hostUsers: false` on Kubernetes 1.27 with CRI-O 1.26 / 1.27.

### Anything else we need to know?

Downgrading Kubelet to 1.26.3 solves this issue.

### CRI-O and Kubernetes version

<details>

```console
$ crio --version
crio version 1.27.0
Version:        1.27.0
GitCommit:      unknown
GitCommitDate:  unknown
GitTreeState:   clean
BuildDate:      2023-04-19T12:36:23Z
GoVersion:      go1.20.3
Compiler:       gc
Platform:       linux/amd64
Linkmode:       dynamic
BuildTags:
  containers_image_ostree_stub
  apparmor
  seccomp
LDFlags:           -X github.com/cri-o/cri-o/internal/pkg/criocli.DefaultsPath="" -X github.com/cri-o/cri-o/internal/version.buildDate=2023-04-19T12:36:23Z -compressdwarf=false -linkmode external
SeccompEnabled:   true
AppArmorEnabled:  true
```

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.1", GitCommit:"4c9411232e10168d7b050c49a1b59f6df9d7ea4b", GitTreeState:"archive", BuildDate:"2023-04-15T11:34:19Z", GoVersion:"go1.20.3", Compiler:"gc", Platform:"linux/amd64"}
```

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
NAME="Arch Linux"
PRETTY_NAME="Arch Linux"
ID=arch
BUILD_ID=rolling
ANSI_COLOR="38;2;23;147;209"
HOME_URL="https://archlinux.org/"
DOCUMENTATION_URL="https://wiki.archlinux.org/"
SUPPORT_URL="https://bbs.archlinux.org/"
BUG_REPORT_URL="https://bugs.archlinux.org/"
PRIVACY_POLICY_URL="https://terms.archlinux.org/docs/privacy-policy/"
LOGO=archlinux-logo
$ uname -a
Linux syquel.de 6.2.11-arch1-1 #1 SMP PREEMPT_DYNAMIC Thu, 13 Apr 2023 16:59:24 +0000 x86_64 GNU/Linux
```

</details>

### Additional environment details (AWS, VirtualBox, physical, etc.)

<details>

</details>

### PR DESCRIPTION
**Pull request by Syquel**
#### What this PR does / why we need it:
Partially fixes #6833 .

#### Which issue(s) this PR fixes:
#6833 

#### Does this PR introduce a user-facing change?

```release-note
fix a segfault when a user namespace is requested
```

### CODE REVIEW

**Comment by haircommander** [comment]
/ok-to-test
/approve

LGTM, thanks @Syquel

**Comment by haircommander** [comment]
PTAL @cri-o/cri-o-maintainers

**Review by giuseppe — APPROVED** [review]
LGTM

File: server/container_create_linux.go [hunk 1]
```diff
@@ -1079,12 +1079,12 @@ func getOCIMappings(m []*types.IDMapping) []rspec.LinuxIDMapping {
 		return nil
 	}
 	ids := make([]rspec.LinuxIDMapping, 0, len(m))
-	for i, m := range m {
-		ids[i] = rspec.LinuxIDMapping{
+	for _, m := range m {
+		ids = append(ids, rspec.LinuxIDMapping{
```
Comment (kolyshkin) [inline server/container_create_linux.go:1083]: Would it be easier to do this change instead?
```diff
-	ids := make([]rspec.LinuxIDMapping, 0, len(m))
+	ids := make([]rspec.LinuxIDMapping, len(m))
```

Comment (Syquel) [reply server/container_create_linux.go:1083]: As far as I understand the intent of the original implementer of this function: `make([]rspec.LinuxIDMapping, len(m))` would already initialize the elements of the underlying array and be therefore slightly less efficient than `make([]rspec.LinuxIDMapping, 0, len(m))`.  

The resulting `ids` should be exactly the same.

**Review by giuseppe — APPROVED** [review]

**Comment by Syquel** [comment]
I can't figure out how to see the error output of the step so I can possibly fix it.

Could someone give me a pointer please?

**Comment by haircommander** [comment]
I think the issue was https://github.com/cri-o/cri-o/issues/6857.

for future reference: 
given a [page](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/cri-o_cri-o/6844/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1653108095611572224)
click [artifacts](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/cri-o_cri-o/6844/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1653108095611572224/)->[artifacts/](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/cri-o_cri-o/6844/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1653108095611572224/artifacts/) -> test name (like [cgroupv2-integration](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/cri-o_cri-o/6844/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1653108095611572224/artifacts/cgroupv2-integration/) -> [cri-o-gather](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/cri-o_cri-o/6844/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1653108095611572224/artifacts/cgroupv2-integration/cri-o-gather/) -> [artifacts](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/cri-o_cri-o/6844/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1653108095611572224/artifacts/cgroupv2-integration/cri-o-gather/artifacts/) -> [testout.txt](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/cri-o_cri-o/6844/pull-ci-cri-o-cri-o-main-ci-cgroupv2-integration/1653108095611572224/artifacts/cgroupv2-integration/cri-o-gather/artifacts/testout.txt)

**Comment by Syquel** [comment]
Thanks for your help!

**Pull request merged**